torch/nn/parallel/distributed.py

```diff
@@ -2238,3 +2240,22 @@ def _check_reducer_finalized(self):
 
     def _set_sparse_metadata(self, global_unique_ids):
         self.reducer._set_sparse_metadata(global_unique_ids)
+
+    def _update_process_group(self, new_process_group):
+        """
+        Dynamically updates the process group for DDP so that we can shrink/expand DDP
+        world size without having to reinitialize DDP.
+
+        NOTE: If you are using custom communications hooks via, register_comm_hook,
+        you need to update the process groups for those hooks separately.
+        """
+        # Force a rebuild of buckets for a new process group. This ensures all ranks
+        # are synchronized in terms of when they will rebuild buckets and also
+        # re-evaluates previous assumptions of buckets given the world size might have
+        # changed.
+        self._has_rebuilt_buckets = False
+        self.reducer._force_bucket_rebuild()
+
+        if not _rank_not_in_group(new_process_group):
```

Comment:
mby this is good enough for validation.  i'm actually wondering, is it invalid to change worldsize using the api or is that arguably valid?  (i guess we could use it to implement elastic tolerance for a dropped rank)

Reply:
> i'm actually wondering, is it invalid to change worldsize using the api or is that arguably valid? (i guess we could use it to implement elastic tolerance for a dropped rank)

This is actually the primary goal of the API, we want to change the worldsize. If you see the unit test below, we change the world size multiple times and ensure that DDP works appropriately and doesn't incur any compile overhead.